python/ray/data/_internal/planner/repartition.py

```diff
@@ -0,0 +1,36 @@
+from typing import Callable, List, Tuple
+
+from ray.data._internal.execution.interfaces import RefBundle, TaskContext
+from ray.data._internal.planner.exchange.push_based_shuffle_task_scheduler import (
+    PushBasedShuffleTaskScheduler,
+)
+from ray.data._internal.planner.exchange.shuffle_task_spec import ShuffleTaskSpec
+from ray.data._internal.planner.exchange.pull_based_shuffle_task_scheduler import (
+    PullBasedShuffleTaskScheduler,
+)
+from ray.data._internal.stats import StatsDict
+from ray.data.context import DatasetContext
+
+
+def generate_repartition_fn(
+    num_outputs: int,
+    shuffle: bool,
+) -> Callable[[List[RefBundle], TaskContext], Tuple[List[RefBundle], StatsDict]]:
+    """Generate function to randomly shuffle each records of blocks."""
+    # TODO: support non-shuffle repartition as _internal/fast_repartition.py.
+    assert shuffle, "Execution optimizer does not support non-shuffle repartition yet."
+
+    def fn(
+        refs: List[RefBundle],
+        ctx: TaskContext,
+    ) -> Tuple[List[RefBundle], StatsDict]:
+        shuffle_spec = ShuffleTaskSpec(random_shuffle=False)
```

Comment:
https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/stage_impl.py#L32-L62 is the corresponding code path for shuffle repartition.